api/types/types.proto

```diff
@@ -2729,3 +2771,115 @@ message RecoveryCode {
+    string Kind = 2 [ (gogoproto.jsontag) = "kind,omitempty" ];
+
+    // State is the current state of this session.
+    SessionState State = 3 [ (gogoproto.jsontag) = "state,omitempty" ];
+
+    // Created encodes the time at which the session was registered with the auth
+    // server.
+    //
+    // This should match the timestamp in the corresponding `session.create` event.
+    // It's thus up to the tracker creator to set the correct timestamp.
+    google.protobuf.Timestamp Created = 4 [
+        (gogoproto.stdtime) = true,
+        (gogoproto.nullable) = false,
+        (gogoproto.jsontag) = "created,omitempty"
+    ];
+
+    // Expires encodes the time at which this session expires and becomes invalid.
+    google.protobuf.Timestamp Expires = 5 [
+        (gogoproto.stdtime) = true,
+        (gogoproto.nullable) = false,
+        (gogoproto.jsontag) = "expires,omitempty"
+    ];
+
+    // AttachedData is arbitrary attached JSON serialized metadata.
+    string AttachedData = 6 [ (gogoproto.jsontag) = "attached,omitempty" ];
+
+    // Reason is an arbitrary string that may be used to describe the session and/or it's
+    // purpose.
+    string Reason = 7 [ (gogoproto.jsontag) = "reason,omitempty" ];
+
+    // Invited is a list of invited users, this field is interpreted by different
+    // clients on a best-effort basis and used for delivering notifications to invited users.
+    repeated string Invited = 8 [ (gogoproto.jsontag) = "invited,omitempty" ];
+
+    // Hostname identifies the target this session is connected to.
+    string Hostname = 9 [ (gogoproto.jsontag) = "target_hostname,omitempty" ];
+
+    // Address is the address of the target this session is connected to.
+    string Address = 10 [ (gogoproto.jsontag) = "target_address,omitempty" ];
+
+    // ClusterName is the name of the Teleport cluster that this session belongs to.
+    string ClusterName = 11 [ (gogoproto.jsontag) = "cluster_name,omitempty" ];
+
+    // Login is the local login/user on the target used by the session.
+    string Login = 12 [ (gogoproto.jsontag) = "login,omitempty" ];
+
+    // Participants is a list of session participants.
+    repeated Participant Participants = 13
+        [ (gogoproto.jsontag) = "participants,omitempty", (gogoproto.nullable) = false ];
+
+    // The Kubernetes cluster this session belongs to.
+    string KubernetesCluster = 14 [ (gogoproto.jsontag) = "kubernetes_cluster,omitempty" ];
+
+    // HostUser is the user regarded as the owner of this session, RBAC checks are performed
+    // against the require policies of this user.
+    //
+    // This refers to the Teleport user but may not be the same as the sessions initiator.
+    string HostUser = 15 [ (gogoproto.jsontag) = "host_user,omitempty" ];
+
+    // HostRoles is a list of roles held by the host user at the time of session creation.
+    repeated RoleV5 HostRoles = 16 [ (gogoproto.jsontag) = "host_roles,omitempty" ];
```

Comment:
Can we just store the role names and the `SessionRequirePolicy`s here instead?  Still working on wrapping my head around how it all works, but I think those are the only values we care about, right?

Roles may contain sensitive information unrelated to the needs of this specific system.

Reply:
not quite, we do need more information like metadata and versioning among some other things probably that the RBAC evaluator cares about, i've gone ahead and introduced a PolicySet type which contains just the information the evaluator needs from a given role and switched to attaching that instead. funnily enough it also sped up the evaluator a decent bit